### PR TITLE
Automatic Setup of Prometheus on per Chain Basis

### PR DESCRIPTION
**Pull request by ac9-xyz**
support for Prometheus automation

**Needs to be tested at main.yml level rather than playbook level*

### CODE REVIEW

**Comment by ac9-xyz** [comment]
Add comment on `#{{ chain }}` to blockinfile

**Comment by ac9-xyz** [comment]
check prometheus config with prom-tool

**Comment by ac9-xyz** [comment]
if prom-tool failed then revert

**Comment by ac9-xyz** [comment]
notify helper if failed

**Comment by ac9-xyz** [comment]
make another block in the main.yaml that is the same as support_prometheus

File: main.yml [hunk 1]
```diff
@@ -1,12 +1,22 @@
...
   roles:
     - node_install
     - node_initialize
     - node_configure
     - node_launch
     - role: support_sync_snapshot
-      when: snapshot_url is defined
\ No newline at end of file
+      when: snapshot_url is defined
+    - role: support_prometheus
```
Comment (cosmos-spaces) [inline main.yml:13]: you dont need this anymore.

File: main.yml [hunk 2]
```diff
@@ -1,12 +1,22 @@
...
+      when: prometheus is defined and prometheus | bool
+- name: Setup Prometheus Monitoring on Chain
+  hosts: telemetry
+  gather_facts: true
```
Comment (cosmos-spaces) [inline main.yml:17]: do you need to father facts? You will gather facts if your playbook runs tasks based on remote host values.

This module is automatically called by playbooks to gather useful variables about remote hosts.

File: roles/support_prometheus/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
+- name: Write new entry to prom-cosmos.yml
+  blockinfile:
+    path: "/etc/prometheus/{{ cosmos_prom_file }}.yml"
```
Comment (cosmos-spaces) [inline roles/support_prometheus/tasks/main.yml:3]: where is cosmos_prom_file defined? Add it as a default for the playbook role. <role_dir>/defaults/main.yml.

To create defaults, simply add a defaults/main. yml file in your role directory. These variables will have the lowest priority of any variables available, and can be easily overridden by any other variable, including inventory variables.

File: roles/support_prometheus/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      - targets: ['{{ hostvars[item].ansible_host }}:{{ custom_port_prefix }}61']
+        labels:
+          instance: {{ hostvars[item].server_hostname }}
+          chain_id: {{ chain_id }}
+          network: {{ target }}
+          type: {{ hostvars[item].type }}
+  loop: "{{ groups[target] }}"
+  when: groups[target] is defined
+
+- name: Check if Prometheus is responding via cURL
```
Comment (cosmos-spaces) [inline roles/support_prometheus/tasks/main.yml:16]: you are not using curl anymore so update the name.

File: roles/support_prometheus/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+  shell: |
+    promtool check healthy
+  register: promHealthy
+
+- name: Check if Prometheus is responding via cURL
```
Comment (cosmos-spaces) [inline roles/support_prometheus/tasks/main.yml:21]: same as above.

File: support_prometheus.yml [hunk 6]
```diff
@@ -0,0 +1,9 @@
+---
+- name: Setup Prometheus Monitoring on Chain
+  hosts: telemetry
+  gather_facts: true
```
Comment (cosmos-spaces) [inline support_prometheus.yml:4]: do you need to gather facts?

File: main.yml (hunk 1 above)
Comment (ac9-xyz) [reply main.yml:13]: dangit, i forgot!

File: main.yml (hunk 2 above)
Comment (ac9-xyz) [reply main.yml]: I think I do not need to gather facts.

File: roles/support_prometheus/tasks/main.yml (hunk 4 above)
Comment (ac9-xyz) [reply roles/support_prometheus/tasks/main.yml]: done.

File: roles/support_prometheus/tasks/main.yml (hunk 5 above)
Comment (ac9-xyz) [reply roles/support_prometheus/tasks/main.yml]: done.

File: support_prometheus.yml (hunk 6 above)
Comment (ac9-xyz) [reply support_prometheus.yml]: I think I do not need to gather facts.

File: main.yml [hunk 7]
```diff
@@ -1,12 +1,20 @@
...
     - node_configure
     - node_launch
     - role: support_sync_snapshot
-      when: snapshot_url is defined
\ No newline at end of file
+      when: snapshot_url is defined
+
+- name: Setup Prometheus Monitoring on Chain
+  hosts: telemetry
+  vars_files:
```
Comment (cosmos-spaces) [inline main.yml:16]: explicitly set it to gather_facts: false. Default is true.

File: support_prometheus.yml [hunk 8]
```diff
@@ -0,0 +1,8 @@
+---
+- name: Setup Prometheus Monitoring on Chain
+  hosts: telemetry
+  vars_files:
```
Comment (cosmos-spaces) [inline support_prometheus.yml:4]: explicitly set it to gather_facts: false. Default is true.

File: main.yml (hunk 7 above)
Comment (ac9-xyz) [reply main.yml:17]: done.

File: support_prometheus.yml (hunk 8 above)
Comment (ac9-xyz) [reply support_prometheus.yml:5]: done.

**Review by cosmos-spaces — APPROVED** [review]

**Pull request merged**